src/TaggedProviderTrait.php

```diff
@@ -0,0 +1,110 @@
+     * guarantee the presence of the method returned by tagMethod().
+     *
+     * @return string
+     *   The class or interface type that this Provider is for.
+     */
+    abstract protected function eventType() : string;
+
+    /**
+     * Returns the method to call on an event to get its tag.
+     *
+     * The tag is an opaque string that identifies the event
+     * within the scope of its type.  Examples include a form name,
+     * workflow name, or some other user-configured object.
+     *
+     * @return string
+     *   The tag method to call on an event.
+     */
+    abstract protected function tagMethod(): string;
+
+    /**
+     * Returns an ordered iterable of Listeners relevant to this tag.
+     *
+     * Not all Listeners returned by this method will be invoked. Only
+     * those that are type-compatible with the Event being processed
+     * will be.
+     *
+     * @param string $tag
+     *   The tag for which we want relevant Listeners.
+     * @return iterable
+     *   An ordered iterable of candidate Listeners.  The key of each entry
+     *   is the Event type the Listener is for.  (It will also apply if the
+     *   $type is a parent of the Event's type.)  The value is an iterable
+     *   of Listeners that apply to that type and to the tag that was passed.
+     */
+    abstract protected function getListenersForTag(string $tag) : iterable;
+
+    /**
+     * Returns an ordered iterable of Listeners relevant to all tags.
+     *
+     * Not all Listeners returned by this method will be invoked. Only
+     * those that are type-compatible with the Event being processed
+     * will be.
+     *
+     * Note that this method is called after getListenersForTag(), so
+     * tag-specific Listeners will always be invoked first.
+     *
+     * @return iterable
+     *   An ordered iterable of candidate Listeners.  The key of each entry
+     *   is the Event type the Listener is for.  (It will also apply if the
+     *   $type is a parent of the Event's type.)  The value is an iterable
+     *   of Listeners that apply to that type.
+     */
+    abstract protected function getListenersForAllTags() : iterable;
+
+    public function getListenersForEvent(object $event) : iterable
+    {
+        $eventType = $this->eventType();
+        if (!$event instanceof $eventType) {
+            return [];
+        }
+
+        $tag = $event->{$this->tagMethod()}();
+
+        foreach ($this->getListenersForTag($tag) as $type => $listeners) {
+            foreach ($listeners as $listener) {
+                if ($event instanceof $type) {
+                    yield $listener;
+                }
+            }
+        }
+        foreach ($this->getListenersForAllTags() as $type => $listeners) {
+            foreach ($listeners as $listener) {
+                if ($event instanceof $type) {
+                    yield $listener;
+                }
+            }
+        }
```

Comment:
It seems like there's an awful lot of duplication here. What if the `getListenersForTag()` and `getListenersForAllTags()` methods were to accept the event instance, and be expected to yield listeners? This might then look like:

```php
$tag = $event->{$this->tagMethod()}();

yield from $this->getListenersForTag($event, $tag);
yield from $this->getListenersForAllTags($event);
```

Those methods could then determine how they locate appropriate listeners. Thoughts?

Either that, or have a trait that provides a method that expects an iterable and is itself iterable:

```php
private function yieldListenersForEvent(object $event, iterable $listeners) : iterable
{
    foreach ($listeners as $type => $listener) {
        if ($event instanceof $type) {
            yield $listener;
        }
    }
}
```

and then the above becomes:

```php
yield from $this->yieldListenersForEvent($event, $this->getListenersForTag($tag));
yield from $this->yieldListenersForEvent($event, $this->getListenersForAllTags());
```

That latter trait would be quite useful as a generic mechanism for retrieving listeners in a provider; implementations would be able to basically only need to provide registration at that point.

Reply:
Well, if we push too much code off to implementers then this trait doesn't really do anything useful.  Balancing that is my concern.  If they're given the tag and the event...  Is this trait really doing anything useful at all?